daemon: Top-level composition into a hierarchy of cells

joamaki opened the pull request:
First commit cleans up the `cilium-agent objects` output.
Second commit adds `daemon/cmd/cells.go` that expresses the agent as a composition of 3 top-level modules and refactors daemon_test.go and test/controlplane to use the `cmd.ControlPlane`. This removes the last blocker from being able to start implementing more of the agent as cells.

Here is how the output of `cilium-agent objects` looks now. The agent is divided into 3 top-level modules,
infrastructure, control plane and datapath. The infrastructure and datapath modules are meant to contain implementations that talk to real systems, either external or the local kernel. Control plane is meant to contain pure business logic that can be integration tested in non-privileged platform independent tests (e.g. on both macOS and Linux) by mocking the infrastructure and datapath modules. Preferably over time we would have the source code organization reflect this division (if we can stomach the backport pain).

```
Cells:

  Ⓜ️ Cilium Agent:
      Ⓜ️ Infrastructure:
          Ⓜ️ gops:
              ⚙️ gops.GopsConfig:
                  • gops-port[uint16] = 9890
              🛠️ gops.registerGopsHooks (cell.go:38): func(hive.Lifecycle, logrus.FieldLogger, gops.GopsConfig)

          Ⓜ️ k8s-client:
              ⚙️ client.Config:
                  • enable-k8s-api-discovery[bool]  = false
                  • k8s-api-server[string]          = 
                  • k8s-client-burst[int]           = 0
                  • k8s-client-qps[float32]         = 0
                  • k8s-heartbeat-timeout[duration] = 30s
                  • k8s-kubeconfig-path[string]     = 
              🚧 client.newClientset (cell.go:106): func(hive.Lifecycle, logrus.FieldLogger, client.Config) (client.Clientset, error)


      Ⓜ️ Control Plane:
          🚧 node.newLocalNodeStore (local_node_store.go:71): func(node.LocalNodeStoreParams) (node.LocalNodeStore, error)
          Ⓜ️ Daemon:
              🚧 cmd.newDaemonPromise (daemon_main.go:1614): func(cmd.daemonParams) promise.Promise[*cmd.Daemon]
              🛠️ cmd.glob..func1 (daemon_main.go:1601): func(promise.Promise[*cmd.Daemon])


      Ⓜ️ Datapath:
          🚧 cmd.newDatapath (daemon_main.go:1545): func(hive.Lifecycle) datapath.Datapath

```

joamaki committed:
hive: Trim names and types in PrintObject down to pkgName.var

Package names within github.com/cilium/cilium are unique enough that
we can trim it down:

   github.com/cilium/cilium/pkg/k8s/resource.Resource[*github.com/cilium/cilium/.../v1.Service]
-> resource.Resource[*v1.Service]

In the rare cases there is ambiquity (like above, is it corev1 or slim_corev1?), it's easily resolvable
looking at the constructor implementation. Keeping the output human readable is more important.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

squeed commented on daemon/cmd/daemon_main.go:
```diff
@@ -1679,7 +1702,8 @@ func runDaemon(ctx context.Context, cleaner *daemonCleanup, shutdowner hive.Shut
 	restoreComplete := d.initRestore(restoredEndpoints)
-	if wgAgent != nil {
+
+	if wgAgent, ok := d.datapath.WireguardAgent().(*wireguard.Agent); ok && wgAgent != nil {
```
Rather than casting, can you add whatever methods are missing to the interface?

joamaki replied:
That will create an import cycle (ipcache depends on pkg/datapath, Init() takes *IPCache). I'll instead pass *wg.Agent down as a separate param. It'll be a bit ugly for a short while with this "init after datapath restored". Moving datapath related initialization into pkg/datapath is not far off now though. This was done in the "top-level composition" commit.

squeed commented on daemon/cmd/daemon_main.go:
```diff
@@ -1090,17 +1106,6 @@ func initEnv() {
 		lbmap.SizeofSockRevNat6Key+lbmap.SizeofSockRevNat6Value)
 
-	// Prepopulate option.Config with options from CLI.
-	option.Config.Populate(Vp)
```
A bunch of other stuff in `initEnv` relies on option.Config -- is this going to be a problem?

joestringer replied:
As far as I followed, the `init()` function earlier will ensure that `cobra.OnInitialize()` is set up to call `option.Config.Populate(Vp)`, so then when `main()` -> `cmd.Execute()` -> `RootCmd.Execute()` is called, it should call the OnInitialize() functions to populate the config and then subsequently call `RootCmd.Run` -> `runApp()` -> `initEnv()` and retain the same ordering property as today.

Handy link for cross-reference: https://pkg.go.dev/github.com/spf13/cobra#OnInitialize

joamaki replied:
Yep, the order remains the same, I basically just split it so that all commands get `option.Config.Populate` but only root command calls `initEnv`.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1539,30 +1513,104 @@ func (d *Daemon) initKVStore() {
+
+			iptablesManager.Init()
+			return nil
+		}})
```
Now that this is in the `OnStart()`, this logic is actually deferred until later instead of happening in between the node initialization and the wireguard initialization right? I don't know if that's consequential but it wasn't called out in the commit message.

joamaki replied:
Double checked that it's OK and modified commit message.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1362,30 +1360,6 @@ func initEnv() {
-			node.SetIPv4(ip)
-		}
-	}
-
```
Any chance we could make this ordering change before as a separate commit, in case this has an unknown consequence? It would be helpful to be able to git bisect down to this specific change if it triggers an issue. I know it's called out in the commit message, but it's hard to say whether that would be enough to pinpoint the change.

joestringer replied:
For review purposes I ended up doing this myself, which made it a bit easier for me to reason about anyway (no changes, just git commit splitting): https://github.com/joestringer/cilium/pull/new/review/agent-composition

joamaki replied:
Thanks for splitting it up. Reset to your branch.

joamaki replied:
Reverted this change in the end to minimize the number of functional changes in this PR as I hit too many surprises.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1539,30 +1513,104 @@ func (d *Daemon) initKVStore() {
-	if params.Config.SkipDaemon {
-		return nil
-	}
+func newDaemonPromise(params daemonParams) promise.Promise[*Daemon] {
```
Could you paint a bit more of a picture of the rationale behind the promise usage here? I'm sure there's good reasons, but I was not able to parse them from the commit message :-)

Here's the current explanation:

    To facilitate gradual move of components into cells from Daemon struct,
    this expose it as promise.Promise[*Daemon]. This can be then further
    projected down to individual objects as needed. E.g. by adding provider:
    func projectIPCache(promise.Promise[*Daemon]) promise.Promise[*IPCache].

Looking through the commits, I only see `daemonPromise.Await(ctx)` being used in unit tests, so maybe it's making some aspect of dependency injection or synchronization easier? My naive read of the description above is "use promises in the Daemon so that we can use promises in subsequent components".

joestringer replied:
Thinking a little bit more, the overall structure right now is effectively one giant function that triggers all of the initializers across the entire codebase. I *think* the rationale here is about trying to make each individual dependency more explicit, so that over time we can split up the giant function into lots of smaller modules that are defined independently with only the logic relevant to those modules. Then it'll be hopefully easier to reason about initialization ordering for each hunk of logic, since there will only be a list of promises that are required and provided and anything beyond that is not relevant to this module.

joestringer replied:
Second day thoughts on the above is that most dependencies should already be handled by the Hive / Cell design & parameters being automatically resolved. But probably there is more use for promises in cases where a module runs its own dedicated goroutines for module-internal logic, and that logic needs to rely on the availability of some other module. I may have a good use for this in ipcache, since there's currently some ugly logic around doing some initial setup and then retrying the main loop if it gets triggered before identity allocation or k8s is fully synced. Not sure how much of that can be resolved via Cells vs. promises yet, but we can figure that out over time.

joamaki replied:
The main reason for the `Promise[*Daemon]` right now is that we're using it in tests (daemon_test.go, controlplane) to grab `*Daemon` and use it to validate results.

We can also use it to "lift" things out from `*Daemon` and depend on them in other components, but of course preferably we would not depend on `Promise[*Foo]` and would rather provide `*Foo` normally via injection. I'll reword the commits to not suggest this and just document the current use.

joestringer reviewed (commented):
I like the three-layering. I didn't take a closer look at the subsequent PR yet but at a high level it seems like a reasonable step in the right direction.

While reviewing the second commit, I wondered if some of the refactoring / code moves could be split into dedicated commits in order to make the rest of the functional changes a bit more obvious. I know it's not all that much code, for example the factoring out of the datapath initialization into `newDatapath()`, but every line of code that is only moving from one place to another into a dedicated function makes it more difficult to pay attention to the functional changes. I realize the wrapping in the function is also then essential to some of the refactoring/reordering, but if there's a way to e.g. move the code to a function without changing anything else in one commit, then do the functional change separately, that would be easier to review. I noted one other example below with only reordering of some node init logic that might be another candidate for a dedicated commit.

At a glance of the testsuite, it seems like the monitor initialization ordering is being changed. I was scanning through to try to locate why this might be, but it wasn't obvious to me just yet. The reason I suspect this is [in this failure](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.25-kernel-net-next/471/testReport/junit/Suite-k8s-1/25/K8sDatapathConfig_Wireguard_encryption_Pod2pod_is_encrypted_in_tunneling_mode/):

```
⚠️  Number of "level=warning" in logs: 53
...
Top 3 errors/warnings:
Notifying monitor about endpoint regeneration failed
Failed to send agent start monitor message
```

Notably also, other failures in the ginkgo suite appear to be giving a signal that the monitor event processing / output might not be keeping up.

After one pass, I'm not confident I understand the totality of the changes well enough to give an explicit approval, but generally they seem like reasonable changes that are a necessary step on the path to modularizing the agent. :+1: Maybe I'll sleep on it and take another pass.

joamaki committed:
daemon: Move CRD identity allocation mode to NewDaemon

Move the CRD identity allocation mode check from initEnv to NewDaemon as it
checks if k8s-client is enabled, which is not yet configured when initEnv()
is invoked.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joamaki committed:
daemon: Refactor and combine datapath initialization

Move the datapath initialization code to a new dedicated function.

The IptablesManager.Init call is deferred to a start hook to allow
"cilium-agent objects" to work. This does not change the initialization order
as nothing in newDatapath uses iptablesManager and the returned Datapath interface
is only used when initializing *Daemon which happens from a start hook.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joamaki committed:
daemon: Call initEnv only when executing via root command

initEnv() is only relevant when running the agent as root via the cilium-agent
binary and should be skipped in integration tests.

This moves the initEnv call into runApp() from registerDaemonHooks(). Since we do want
option.Config.Populate and logging to be set up for all commands (dot-graph, objects),
these are moved from initEnv() into cobra.OnInitialize hook (executed prior to the command
 run hook).

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Top-level composition into a hierarchy of cells

This is the first step towards composing the agent as a hierarchy of cells
divided into 3 layers: infrastructure, control plane and datapath.

By dividing into these layers we can start having a clearer separation
of responsibilities and allow writing integration tests without having
to use "DryMode" as a crutch. The "DryMode" becomes mocked "Infrastructure"
and "Datapath".

daemon tests and test/controlplane are refactored to run the agent via
cmd.ControlPlane rather than constructing it directly via NewDaemon.

Since the tests use *Daemon to validate, provide it asynchronously via
Promise[*Daemon]. Providing it directly would require splitting NewDaemon
into pure object construction and start hook which is currently non-trivial.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer reviewed (approved):
Overall the changes look good to me. There's several ordering modifications in the last commit, so it'd be nice if we could split each ordering change into a dedicated commit in a similar style to how I did [here](https://github.com/joestringer/cilium/pull/new/review/agent-composition) just to make things a bit more bisectable in case of problems. Other than that, I have a few discussion points above but I don't think that any of these are particularly consequential to the structural changes being proposed in this PR.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1587,76 +1638,48 @@ func registerDaemonHooks(params daemonParams) error {
 			// Start running the daemon in the background (blocks on API server's Serve()) to allow rest
 			// of the start hooks to run.
-			go runDaemon(ctx, cleaner, params.Shutdowner, params.Clientset)
+			if !option.Config.DryMode {
```
Interesting! That means that the other stuff in `runDaemon` is no longer necessary for unit testing :+1: 

I noticed that the `runDaemon()` function still has some `!option.Config.DryMode` checks inside, so we can probably drop those since the check now occurs here?

joamaki replied:
Removed the DryMode checks from runDaemon and initEnv.

joamaki committed:
daemon: Drop now unnecessary DryMode checks

initEnv and runDaemon are now only invoked when not running in DryMode, so
DryMode checks in these can now be removed.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on daemon/cmd/daemon.go:
```diff
@@ -903,10 +910,10 @@ func NewDaemon(ctx context.Context, cleaner *daemonCleanup,
+		if err := wgAgent.Init(d.ipcache, d.mtuConfig); err != nil {
 			log.WithError(err).Error("failed to initialize wireguard agent")
-			return nil, nil, fmt.Errorf("failed to initialize wireguard agent: %w", err)
+			return nil, nil, fmt.Errorf("failed to initilaize wireguard agent: %w", err)
```
Oops typo here somehow.

joamaki committed:
node: Use a singleton LocalNodeStore and don't swap it

Since we're still using the global getters and setters in pkg/node
it's fairly dangerous to swap it out with SetLocalNodeStore as any
code that runs prior to that call and uses pkg/node may end up
modifying state that gets thrown away.

To make the transition safer, revert back to the old semantics and
keep a singleton LocalNodeStore that needs to be reset in tests via
Uninitialize. We'll need to revisit this when the global getters&setters
are removed.

This fixes an issue in a prior commit that moved wireguard agent initialization
earlier than SetLocalNodeStore.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
